library/src/scala/tasty/reflect/Kernel.scala

```diff
@@ -909,33 +910,41 @@ trait Kernel {
 
   def ConstantType_constant(self: ConstantType) given (ctx: Context): Constant
 
-  /** Type of a reference to a symbol */
-  type SymRef <: Type
+  /** Type of a reference to a term symbol */
+  type SymTermRef <: Type
 
-  def matchSymRef(tpe: TypeOrBounds) given (ctx: Context): Option[SymRef]
+  def matchSymTermRef(tpe: TypeOrBounds) given (ctx: Context): Option[SymTermRef]
 
-  // TODO remove this method. May require splitting SymRef into TypeSymRef and TermSymRef
-  def matchSymRef_unapply(tpe: TypeOrBounds) given (ctx: Context): Option[(Symbol, TypeOrBounds /* Type | NoPrefix */)]
+  def matchSymTermRef_unapply(tpe: TypeOrBounds) given (ctx: Context): Option[(Symbol, TypeOrBounds /* Type | NoPrefix */)]
 
-  def SymRef_qualifier(self: SymRef) given (ctx: Context): TypeOrBounds
+  def SymTermRef_qualifier(self: SymTermRef) given (ctx: Context): TypeOrBounds
 
-  /** Type of a reference to a term */
-  type TermRef <: Type
+    /** Type of a reference to a type symbol */
+  type SymTypeRef <: Type
```

Comment:
Is `SymTypeRef` a subtype of `NameTypeRef`? Meanwhile, if most of the time it's possible to get a symbol from `NamedTypeRef` (of course not for type member of structural types), then the subtle difference of the two concepts seems not helpful to meta-programmers, only causing confusions.

The concern above is purely from the viewpoint of meta-programmers. I'm not familiar enough to tell from the perspective of compiler construction. 

/cc: @odersky

Reply:
Those are not subtypes of each other. The current version caused confusion on you and @milessabin as the expectation was that a TypeRef would match both a type ref with a named or symbol designator. This change makes it explicit that type references could one of the two and they have to be handled slightly differently.

We could also add more subtyping by adding the generic TypeRef and TermRef.
```scala
type Type
  type TermRef <: Type
      type NamedTermRef <: Type
      type SymTermRef <: Type
  type TypeRef <: Type
      type NamedTypeRef <: Type
      type SymTypeRef <: Type
```
I did not add this into this PR because it make the hierarchy more complex.

Reply:
I'm a bit confused about how this relates to the Dotty internals. I quite like the internal `TypeRef`/`TermRef` (both a subtype of `NamedType`) distinction along with `typeSymbol` and `termSymbol`. Isn't it the case that if something has a name then it has a symbol? In which case I don't really understand why the `TypeRef`/`TermRef` extractors give us the name as a `String` and then makes us jump through hoops to get the symbol.

Reply:
If I remember correctly, you can have type refinements where you have a name but no symbol.

Reply:
Makes sense. But surely that's the (much) less common case ... why not handle it separately?

Reply:
The latest change uses `TypeRef`/`TermRef` for the ones with symbols and renames the named ones to `NamedTypeRef`/`NamedTermRef`. This will make the common simpler to match on.